### DISCUSSION TITLE
Next.js Downloadable Files for Digital Goods

### DISCUSSION THREAD

**Discussion by surjithctly** [vercel/next.js#17223] [Help]
Hey, First time here, Just wanted to see if Next JS fits for my needs. 

Suppose, I have a digital goods marketplace, I want to build it with Next JS, I can handle payments via Stripe. 

Now I wanted an option to provide users to download the "purchased digital goods" via Next JS. 

So, only purchased users will able to access the file (check with stripe) . Otherwise gets an error. I will be hosting it in Vercel. 

So, Is it possible to restrict accessing files only for authenticated users? 

Let me know the details. 

Thanks so much !

~Surjith

**Comment by leerob** [comment]
Yes, absolutely! There are quite a few different ways you can do this. 

1. Client-side application, using a React hook to check for a user (e.g. `useUser`)
2. Server-rendered application, checking for a user via a cookie / JWT in `getServerSideProps`

I'm assuming your application will allow users to log in with some provider (Firebase, Auth0, etc). There are a few different examples in the Next.js repo for these. Let me know if you have any more questions. 

https://github.com/vercel/next.js/tree/canary/examples

**Comment by surjithctly** [comment]
@leerob Thanks a lot for replying.

My doubt is where do I put these files securely. I heard usually such assets are stored in /public folder. But I worry everyone will be able to download the file.

If you are doing such app, where will you put the files (eg: zip) and how you make it download only if authentication + purchase verification done? Yet also hide it from general public.

Thanks again.

**Reply by weyert** [reply]
I would suggest something like blob storage, like Azure Storage, Google Storage or S3 and then set the permissions correctly so they can’t be accessed by everyone. Next, you generate a signed url for the authenticated user, which secures it, and allows you to download it.

**Reply by surjithctly** [reply]
That’s good idea.

But does that also means its not possible with storing files locally?

**Comment by Manoj-kumar-C** [comment]
Really nice
